### ISSUE TITLE
[bug] BG2:  Known spells reverse their order on save-load

### ISSUE THREAD

**Issue by caillean7** [gemrb/gemrb#324]
Split from #171

**Description**

> known spells reverse their order on save-load

I'm pretty sure I can remember that from a couple of years ago, but can't reproduce anymore. Might be already taken care of?

**Comment by lynxlynxlynx** [comment]
Trivial to reproduce with a mage. Remember, it's the known spells, not the memorized ones.

**Comment by ifyGecko** [comment]
The known spell list reverses on save. Continually re-loading the same save does not reverse the order.

**Comment by lynxlynxlynx** [comment]
Yeah, most likely just the iteration when saving has to be reversed.

**Comment by Tomsod** [comment]
Continuously re-loading can't give different results either way, because you're repeating the same deterministic operation.  It would probably be more prudent to test (1) saving in IE and loading in GemRB, and (2) saving in GemRB and loading in IE.  If (1) reverses spells, fix loading; if (2) does, fix saving.  That way the fix will be compatible with IE.

**Comment by caillean7** [comment]
I just tried real quick:

> (1) saving in IE and loading in GemRB

Load order is fine in GemRB (ascending alphabetically)

> (2) saving in GemRB and loading in IE

A save from GemRB reloaded in GemRB has the order reversed, while loading the same in IE has the right load order.

**Comment by Tomsod** [comment]
Tried the same, and I think I get it.  Indeed, loading a save in (original) BG2 sorts the spells in alphabetical order!  I never noticed that before.  It respects localization, as well (my spells were sorted according to their Russian names), so the order must be computed on the fly.

Meanwhile, in GemRB learning a new spell adds it to the end of the list, and the list is also flipped on each save/load.  If you're testing with an IE save, you won't notice, but I played through BG1 in GemRB and my spellbook is a mess.

Also, saving the game in IE and then loading it in GemRB preserves the alphabetical ordering.  That means it's not merely a display thing and IE actually overwrites the spell order in the save.  Also, it probably means that GemRB reverses spells on save, not load (otherwise it would load spells in Z-A order).  But I think we should just drop the whole reversal issue and simulate IE behavior instead.

**Comment by caillean7** [comment]
It certainly looks as if IE does the sorting on both saving and loading.

**Comment by lynxlynxlynx** [comment]
@ifyGecko proposed the changes to maintain the order, but thanks for the extra research. Anything that shortens iediff output when comparing CREs is welcome. I doubt anything would break with this, but I'm curious if bg1 also did the same reordering.

**Comment by caillean7** [comment]
Not on save game loading. I tried loading one from GemRB with reversed order, it showed up in the original also reversed.

**Comment by Tomsod** [comment]
If we are doing this, we should also handle inserting a newly learned spell so as to preserve the alphabetical order.  Assuming IE does that too.

**Comment by lynxlynxlynx** [comment]
I'm closing this and opening a new issue for the order, so the discussion can be more focused.
